janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrResultIterator.java

```diff
@@ -0,0 +1,106 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.diskstorage.solr;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.function.Function;
+
+import org.apache.solr.client.solrj.SolrClient;
+import org.apache.solr.client.solrj.SolrQuery;
+import org.apache.solr.client.solrj.SolrServerException;
+import org.apache.solr.client.solrj.StreamingResponseCallback;
+import org.apache.solr.common.SolrDocument;
+import org.janusgraph.diskstorage.PermanentBackendRuntimeException;
+
+/**
+ * @author David Clement (david.clement90@laposte.net)
+ */
+class SolrResultIterator<E> implements Iterator<E> {
+
+    private final SolrClient solrClient;
+    private final BlockingQueue<E> queue;
+    private int cpt;
+    private int nbRequest;
+    private final Long limit;
+    private final int offset;
+    private final int nbDocByQuery;
+    private final String collection;
+    private final SolrQuery solrQuery;
+    private final Function<SolrDocument, E> function;
+
+    public SolrResultIterator(SolrClient solrClient, Integer limit, int offset, int nbDocByQuery, String collection, SolrQuery solrQuery, Function<SolrDocument, E> function) throws SolrServerException, IOException {
+        this.solrClient = solrClient;
+        cpt = 0;
+        this.offset = offset;
+        this.nbDocByQuery = nbDocByQuery;
+        queue = new LinkedBlockingQueue<>();
+        this.collection = collection;
+        this.solrQuery = solrQuery;
+        this.function = function;
+        long nbFound = solrClient.queryAndStreamResponse(collection, solrQuery, new SolrCallbackHandler(this, function)).getResults().getNumFound() - offset;
+        this.limit = limit != null ? Math.min(nbFound, limit) : nbFound;
+        nbRequest = 1;
+    }
+
+    public BlockingQueue<E> getQueue(){
+        return queue;
+    }
+
+    @Override
+    public boolean hasNext() {
+        if(cpt!=0 && cpt % nbDocByQuery == 0 && nbDocByQuery * nbRequest < limit){
+            try {
+                solrQuery.setStart(nbRequest * nbDocByQuery + offset);
+                solrClient.queryAndStreamResponse(collection, solrQuery, new SolrCallbackHandler(this, function));
+                nbRequest++;
+            } catch (SolrServerException | IOException e) {
+                throw new PermanentBackendRuntimeException(e);
```

Comment:
Add a new `UncheckedSolrException` class for wrapping the `SolrServerException`, then break this into two catch blocks to throw UncheckedSolrException or UncheckedIOException, respectively.